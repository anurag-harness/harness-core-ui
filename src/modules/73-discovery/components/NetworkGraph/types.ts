/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { LayoutOptions } from 'elkjs/lib/elk.bundled'
import type { Node, Edge } from 'reactflow'
import type {
  ApiCustomServiceConnection,
  DatabaseConnection,
  DatabaseDiscoveredServiceCollection,
  DatabaseNetworkMapEntity
} from 'services/servicediscovery'

export interface LayoutedGraph {
  nodes: Node<unknown, string | undefined>[]
  edges: Edge<EdgeData>[]
  options?: LayoutOptions
}

export interface EdgeData {
  parentNode?: string
}

export type ServiceNodeData = Partial<DatabaseDiscoveredServiceCollection> & { id: string; name: string }
export type ServiceEdgeData = EdgeData & ApiCustomServiceConnection

export type NetworkMapNodeData = DatabaseNetworkMapEntity
export type NetworkMapEdgeData = EdgeData & DatabaseConnection
