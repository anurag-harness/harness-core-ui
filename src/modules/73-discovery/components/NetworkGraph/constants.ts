/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { ConnectionLineType, Edge, Node, MarkerType } from 'reactflow'

export enum NodeTypes {
  Hexagon = 'hexagon',
  Group = 'group',
  NetworkMapHexagon = 'networkMapHexagon'
}

export const connectionLineType = ConnectionLineType.Straight

export const edgeOptions: Partial<Edge> = {
  animated: true,
  type: connectionLineType,
  markerEnd: { type: MarkerType.ArrowClosed }
}

export const nodeOptions: Omit<Node<unknown, NodeTypes>, 'id' | 'data'> = {
  position: { x: 0, y: 0 },
  width: 150,
  height: 150,
  type: NodeTypes.Hexagon,
  expandParent: true
}

export const nodeGroupOptions: Omit<Node<unknown, NodeTypes>, 'id' | 'data'> = {
  position: { x: 0, y: 0 },
  type: NodeTypes.Group,
  style: {
    border: '2px dashed var(--purple-600)',
    background: 'rgba(246, 241, 255, 0.30)'
  }
}
