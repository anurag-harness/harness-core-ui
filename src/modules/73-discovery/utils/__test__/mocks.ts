/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { DatabaseNetworkMapEntity } from 'services/servicediscovery'

export const connectionsBetweenServices = [
  {
    from: {
      id: '64920dc166c663ba792cf3b0',
      kind: 'discoveredservice',
      name: 'access-control',
      kubernetes: { namespace: 'chaos-1000' }
    },
    port: '9090',
    to: {
      id: '64920dc166c663ba792cf3bb',
      kind: 'discoveredservice',
      name: 'harness-manager',

      kubernetes: { namespace: 'chaos-1000' }
    },
    type: 'TCP',
    manual: false
  },
  {
    from: {
      id: '64920dc166c663ba792cf3bb',
      kind: 'discoveredservice',
      name: 'access-control',

      kubernetes: { namespace: 'chaos-1000' }
    },
    port: '6379',
    to: {
      id: '64920dc166c663ba792cf3b0',
      kind: 'discoveredservice',
      name: 'redis-sentinel-harness-announce-0',

      kubernetes: { namespace: 'chaos-1000' }
    },
    type: 'TCP',
    manual: false
  }
]

export const mockConnections = {
  items: [
    {
      id: '64920e3866c663ba792cf569',
      type: 'TCP',
      sourceID: '64920dc166c663ba792cf3b0',
      sourceName: 'access-control',
      sourceNamespace: 'chaos-1000',
      sourceIP: '10.100.2.2',
      destinationID: '64920dc166c663ba792cf3bb',
      destinationName: 'harness-manager',
      destinationNamespace: 'chaos-1000',
      destinationIP: '10.104.11.119',
      destinationPort: '9090'
    },
    {
      id: '64920e3866c663ba792cf56a',
      type: 'TCP',
      sourceID: '64920dc166c663ba792cf3bb',
      sourceName: 'access-control',
      sourceNamespace: 'chaos-1000',
      sourceIP: '10.100.2.2',
      destinationID: '64920dc166c663ba792cf3b0',
      destinationName: 'redis-sentinel-harness-announce-0',
      destinationNamespace: 'chaos-1000',
      destinationIP: '10.104.0.162',
      destinationPort: '6379'
    }
  ]
}

export const mockServices = {
  items: [
    {
      id: '64920dc166c663ba792cf3b0',
      agentID: '64920dbb66c663ba792cf134',
      kind: 'Service',
      name: 'access-control',
      namespace: 'chaos-1000',
      uid: '02bfaa4d-9fcd-49ed-9e1a-08a24c491c89',
      service: {
        owner: {
          kind: 'Service',
          namespace: 'chaos-1000',
          name: 'access-control',
          uid: '02bfaa4d-9fcd-49ed-9e1a-08a24c491c89',
          apiVersion: 'v1'
        },
        ports: [
          {
            name: 'http',
            protocol: 'TCP',
            port: 9006,
            targetPort: 9006
          }
        ],
        clusterIP: '10.104.11.160',
        clusterIPs: ['10.104.11.160'],
        externalIPs: undefined,
        loadBalancerIP: '',
        externalName: '',
        type: 'ClusterIP'
      },
      createdAt: '2023-06-20T20:36:17.94Z',
      updatedAt: '2023-06-20T20:36:18.357Z',
      createdBy: '',
      updatedBy: '',
      removed: false
    },
    {
      id: '64920dc166c663ba792cf3bb',
      agentID: '64920dbb66c663ba792cf134',
      kind: 'Service',
      name: 'chaos-exporter',
      namespace: 'chaos-1000',
      uid: '02bfaa4d-9fcd-49ed-9e1a-08a24c491c88',
      service: {
        owner: {
          kind: 'Service',
          namespace: 'chaos-1000',
          name: 'chaos-exporter',
          uid: '02bfaa4d-9fcd-49ed-9e1a-08a24c491c89',
          apiVersion: 'v1'
        },
        ports: [
          {
            name: 'http',
            protocol: 'TCP',
            port: 9006,
            targetPort: 9006
          }
        ],
        clusterIP: '10.104.11.160',
        clusterIPs: ['10.104.11.160'],
        externalIPs: undefined,
        loadBalancerIP: '',
        externalName: '',
        type: 'ClusterIP'
      },
      createdAt: '2023-06-20T20:36:17.94Z',
      updatedAt: '2023-06-20T20:36:18.357Z',
      createdBy: '',
      updatedBy: '',
      removed: false
    }
  ],
  page: {
    all: true,
    index: 0,
    totalItems: 1,
    totalPages: 1
  }
}

export const mockNetworkMapResources: DatabaseNetworkMapEntity[] = [
  {
    id: '64920dc166c663ba792cf3b0',
    kind: 'discoveredservice',
    name: 'access-control',
    kubernetes: {
      namespace: 'chaos-1000'
    }
  },
  {
    id: '64920dc166c663ba792cf3bb',
    kind: 'discoveredservice',
    name: 'chaos-exporter',
    kubernetes: {
      namespace: 'chaos-1000'
    }
  }
]

export const mockLogData = `{"level":"","pos":1,"out":"{file:entry.go:41,func:github.com/wings-software/service-discovery/collector.Start,level:info,msg:go version go1.20.8,time:2023-10-04T04:30:15.872344691Z}","time":"2023-10-04T04:30:20.480817808Z","args":{"PodName":"sd-cluster-9ts9x","PodNamespace":"hce"}}
{"level":"","pos":2,"out":"{file:entry.go:42,func:github.com/wings-software/service-discovery/collector.Start,level:info,msg:go os linux,time:2023-10-04T04:30:15.872758918Z}","time":"2023-10-04T04:30:20.843837081Z","args":{"PodName":"sd-cluster-9ts9x","PodNamespace":"hce"}}
{"level":"","pos":2,"out":"{file:entry.go:42,func:github.com/wings-software/service-discovery/collector.Start,level:info,msg:go os linux,time:2023-10-04T04:30:15.872758918Z}","time":"2023-10-04T04:30:20.843837081Z","args":{"PodName":"sd-node-2gc9l","PodNamespace":"hce"}}`

export const mockLogDataResponse = [
  {
    text: {
      level: '',
      out: '{file:entry.go:41,func:github.com/wings-software/service-discovery/collector.Start,level:info,msg:go version go1.20.8,time:2023-10-04T04:30:15.872344691Z}',
      args: {
        PodName: 'sd-cluster-9ts9x',
        PodNamespace: 'hce'
      }
    }
  },
  {
    text: {
      level: '',
      out: '{file:entry.go:42,func:github.com/wings-software/service-discovery/collector.Start,level:info,msg:go os linux,time:2023-10-04T04:30:15.872758918Z}',
      args: {
        PodName: 'sd-cluster-9ts9x',
        PodNamespace: 'hce'
      }
    }
  },
  {
    text: {
      args: {
        PodName: 'sd-node-2gc9l',
        PodNamespace: 'hce'
      },
      level: '',
      out: '{file:entry.go:42,func:github.com/wings-software/service-discovery/collector.Start,level:info,msg:go os linux,time:2023-10-04T04:30:15.872758918Z}'
    }
  }
]

export const installations = {
  id: '651cea4d2dc5d1dc8a74530b',
  agentID: '65126c42183f27a2381f9384',
  delegateTaskID: '2c4sK0KkRsivFp6gb4kDRA-DEL',
  delegateID: '',
  delegateTaskStatus: 'SUCCESS',
  agentDetails: {
    cluster: {
      name: 'sd-cluster-9ts9x',
      namespace: 'hce',
      uid: 'c63fbf54-0f32-4f3a-ad54-078cf8bddf41',
      status: 'Succeeded'
    },
    node: [
      {
        name: 'sd-node-2gc9l',
        namespace: 'hce',
        uid: '234d86b8-454d-434f-87ff-5493e0f0f1ea',
        status: 'Succeeded'
      }
    ],
    status: 'SUCCESS'
  },
  isCronTriggered: true,
  logStreamID: 'c70fbd79-4f7d-45e7-b65c-df4b6237336f',
  logStreamCreatedAt: '2023-10-04T04:30:05.339Z',
  isLogStreamOpen: false,
  createdAt: '2023-10-04T04:30:05.339Z',
  updatedAt: '2023-10-04T04:45:15.299Z',
  createdBy: '',
  updatedBy: '',
  removed: false
}
