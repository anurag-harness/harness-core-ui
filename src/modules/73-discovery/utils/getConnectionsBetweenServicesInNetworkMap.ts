/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import {
  ApiListDiscoveredServiceConnection,
  DatabaseConnection,
  DatabaseConnectionType,
  DatabaseNetworkMapEntity
} from 'services/servicediscovery'

export default function getConnectionsBetweenServicesInNetworkMap(
  services: DatabaseNetworkMapEntity[] | null | undefined,
  connectionList: ApiListDiscoveredServiceConnection | null
): DatabaseConnection[] | undefined {
  // Checks for new connections between the selected services
  const serviceHashMap = new Map<string, DatabaseNetworkMapEntity>(services?.map(s => [s.id ?? '', s]))

  const connectionBetweenSelectedServices: DatabaseConnection[] | undefined = connectionList?.items
    ?.filter(conn => serviceHashMap.has(conn.sourceID ?? '') && serviceHashMap.has(conn.destinationID ?? ''))
    .map(conn => ({
      from: {
        id: conn.sourceID,
        kind: serviceHashMap.get(conn.sourceID ?? '')?.kind ?? 'discoveredservice',
        name: conn.sourceName,
        kubernetes: {
          namespace: conn.sourceNamespace
        }
      },
      port: conn.destinationPort,
      to: {
        id: conn.destinationID,
        kind: serviceHashMap.get(conn.destinationID ?? '')?.kind ?? 'discoveredservice',
        name: conn.destinationName,
        kubernetes: {
          namespace: conn.destinationNamespace
        }
      },
      type: conn.type as DatabaseConnectionType,
      manual: false
    }))

  return connectionBetweenSelectedServices
}
