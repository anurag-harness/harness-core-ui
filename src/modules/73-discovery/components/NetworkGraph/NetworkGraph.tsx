/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import ReactFlow, {
  addEdge,
  useNodesState,
  useEdgesState,
  useReactFlow,
  Controls,
  MiniMap,
  Background,
  BackgroundVariant,
  Node
} from 'reactflow'

import 'reactflow/dist/style.css'
import { getLayoutedElements } from './utils/getLayoutedElements'
import HexagonNode from './components/nodes/HexagonNode/HexagonNode'
import { connectionLineType, edgeOptions } from './constants'
import type { LayoutedGraph } from './types'
import NamespaceGroupNode from './components/nodes/NamspaceGroupNode/NamespaceGroupNode'
import NetworkMapHexagonNode from './components/nodes/NetworkMapHexagonNode/NetworkMapHexagonNode'

export type NetworkGraphProps = Omit<LayoutedGraph, 'options'> & {
  isNodeConnectable?: boolean
  onNodeConnection?: (source: string, target: string) => void
  onNodeClick?: (node: Node<any, string | undefined>) => void
}

const nodeTypes = {
  hexagon: HexagonNode,
  group: NamespaceGroupNode,
  networkMapHexagon: NetworkMapHexagonNode
}

const proOptions = { hideAttribution: true }

export default function NetworkGraph({
  nodes: initialNodes,
  edges: initialEdges,
  isNodeConnectable = false,
  onNodeConnection,
  onNodeClick
}: NetworkGraphProps): React.ReactElement {
  const [nodes, setNodes, onNodesChange] = useNodesState([])
  const [edges, setEdges, onEdgesChange] = useEdgesState([])
  const { fitView } = useReactFlow()

  // This line is being ignored as reactflow doesn't actually render the nodes and edges
  // in a test environment https://github.com/wbkd/react-flow/issues/716.
  // Due to his the user connections with mouse events can't be tested.
  /* istanbul ignore next */
  const onConnect = React.useCallback(
    params => {
      onNodeConnection?.(params.source, params.target)
      setEdges(eds => addEdge({ ...params }, eds))
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [initialNodes, initialEdges]
  )
  const onLayout = React.useCallback(
    // These lines are being ignored as reactflow doesn't actually render the nodes and edges
    // in a test environment https://github.com/wbkd/react-flow/issues/716.
    // Because of this the nodes and edges after the initial render can't be mocked.
    /* istanbul ignore next */
    ({ useInitialNodes = false }) => {
      const ns = useInitialNodes ? initialNodes : nodes
      const es = useInitialNodes ? initialEdges : edges

      getLayoutedElements({ nodes: ns, edges: es }).then(({ nodes: layoutedNodes, edges: layoutedEdges }) => {
        setNodes(layoutedNodes)
        setEdges(layoutedEdges)

        window.requestAnimationFrame(() => fitView())
      })
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [initialNodes, initialEdges, nodes, edges]
  )

  // Calculate the initial layout on mount.
  React.useLayoutEffect(() => {
    onLayout({ useInitialNodes: true })
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [initialNodes, initialEdges])

  return (
    <ReactFlow
      nodes={nodes}
      edges={edges}
      nodeTypes={nodeTypes}
      nodesConnectable={isNodeConnectable}
      onConnect={onConnect}
      onNodesChange={onNodesChange}
      onEdgesChange={onEdgesChange}
      connectionLineType={connectionLineType}
      defaultEdgeOptions={edgeOptions}
      proOptions={proOptions}
      // These lines are being ignored as reactflow doesn't actually render the nodes and edges
      // in a test environment https://github.com/wbkd/react-flow/issues/716.
      // Because of this the nodes and edges after the initial render can't be mocked.
      onNodeClick={/* istanbul ignore next */ (_, node) => onNodeClick?.(node)}
      fitView
    >
      <Controls showInteractive={false} />
      <MiniMap zoomable pannable />
      <Background variant={BackgroundVariant.Dots} color="#aaa" gap={16} size={1} />
    </ReactFlow>
  )
}
