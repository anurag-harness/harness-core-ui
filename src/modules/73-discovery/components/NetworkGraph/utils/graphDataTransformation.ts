/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { Node, Edge } from 'reactflow'
import { NodeTypes, nodeGroupOptions, nodeOptions } from '@discovery/components/NetworkGraph/constants'
import type {
  ApiCreateNetworkMapRequest,
  ApiListDiscoveredServiceConnection,
  ApiListDiscoveredService
} from 'services/servicediscovery'
import type {
  NetworkMapEdgeData,
  NetworkMapNodeData,
  ServiceEdgeData,
  ServiceNodeData
} from '@discovery/components/NetworkGraph/types'

export function getGraphNodesFromServiceList(
  serviceList: ApiListDiscoveredService | null
): Node<ServiceNodeData, NodeTypes>[] {
  if (!serviceList) return []

  const namespaces = new Set<string>()
  const graphNodes: Node<ServiceNodeData, NodeTypes>[] = []
  serviceList.items?.map(service => {
    const namespace = service.spec.kubernetes?.namespace
    if (namespace && !namespaces.has(namespace)) {
      namespaces.add(namespace)
    }

    if (service.id)
      graphNodes.push({
        ...nodeOptions,
        id: service.id,
        data: { ...service, name: service.spec.kubernetes?.name ?? 'service' },
        parentNode: namespace
      })
  })

  namespaces.forEach(value => {
    graphNodes.push({
      ...nodeGroupOptions,
      id: value,
      data: { id: value, name: value }
    })
  })

  return graphNodes
}

export function getGraphEdgesFromServiceConnections(
  connectionList: ApiListDiscoveredServiceConnection | null
): Edge<ServiceEdgeData>[] {
  if (!connectionList) return []

  const graphEdges: Edge<ServiceEdgeData>[] = []

  connectionList.items?.map(connection => {
    if (connection.id && connection.sourceID && connection.destinationID)
      graphEdges.push({
        id: connection.id,
        source: connection.sourceID,
        target: connection.destinationID,
        data: { parentNode: connection.sourceNamespace, ...connection }
      })
  })

  return graphEdges
}

export function getGraphNodesFromNetworkMap(
  networkMap: ApiCreateNetworkMapRequest | undefined
): Node<NetworkMapNodeData, NodeTypes>[] {
  if (!networkMap) return []

  const namespaces = new Set<string>()
  const graphNodes: Node<NetworkMapNodeData, NodeTypes>[] = []

  networkMap.resources?.map(service => {
    const namespace = service.kubernetes?.namespace
    if (namespace && !namespaces.has(namespace)) {
      namespaces.add(namespace)
    }

    if (service.id)
      graphNodes.push({
        ...nodeOptions,
        type: NodeTypes.NetworkMapHexagon,
        id: service.id,
        data: service,
        parentNode: namespace
      })
  })

  namespaces.forEach(value => {
    graphNodes.push({
      ...nodeGroupOptions,
      id: value,
      data: { id: value, name: value, kind: 'discoveredservice' }
    })
  })

  return graphNodes
}

export function getGraphEdgesFromNetworkMap(
  networkMap: ApiCreateNetworkMapRequest | undefined
): Edge<NetworkMapEdgeData>[] {
  if (!networkMap) return []

  const graphEdges: Edge<NetworkMapEdgeData>[] = []

  networkMap.connections?.map(connection => {
    if (connection.from?.id && connection.to?.id)
      graphEdges.push({
        id: `${connection.from.id}-${connection.to.id}`,
        source: connection.from.id,
        target: connection.to.id,
        data: { ...connection, parentNode: connection.from.kubernetes?.namespace }
      })
  })

  return graphEdges
}
