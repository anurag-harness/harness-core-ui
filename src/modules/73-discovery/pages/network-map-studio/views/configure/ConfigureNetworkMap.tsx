/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { Button, ButtonVariation, Container, Layout, Text } from '@harness/uicore'
import { ReactFlowProvider } from 'reactflow'
import { Color, FontVariation } from '@harness/design-system'
import { StudioTabs } from '@discovery/interface/networkMapStudio'
import { ApiCreateNetworkMapRequest, DatabaseNetworkMapEntity } from 'services/servicediscovery'
import { useStrings } from 'framework/strings'
import {
  getGraphEdgesFromNetworkMap,
  getGraphNodesFromNetworkMap
} from '@discovery/components/NetworkGraph/utils/graphDataTransformation'
import NetworkGraph from '@discovery/components/NetworkGraph/NetworkGraph'
import noServiceIllustration from '../../images/noServiceIllustration.png'
import NoSelectionSideBar from './NoSelectionSideBar'
import NodeConnectionsSideBar from './NodeConnectionsSideBar'
import { ServiceRelationFormType } from './RelationPopover'
import css from './ConfigureNetworkMap.module.scss'

export interface ConfigureNetworkMapProps {
  networkMap: ApiCreateNetworkMapRequest
  updateNetworkMap: (networkMap: ApiCreateNetworkMapRequest) => Promise<void>
  handleTabChange: (tabID: StudioTabs) => void
}

export default function ConfigureNetworkMap({
  networkMap,
  updateNetworkMap,
  handleTabChange
}: ConfigureNetworkMapProps): React.ReactElement {
  const { getString } = useStrings()
  const [relation, setRelation] = React.useState<ServiceRelationFormType>()
  const [sourceService, setSourceService] = React.useState<DatabaseNetworkMapEntity>()

  const nodes = React.useMemo(() => getGraphNodesFromNetworkMap(networkMap), [networkMap])
  const edges = React.useMemo(() => getGraphEdgesFromNetworkMap(networkMap), [networkMap])

  function onNodeConnection(sourceId: string, destinationId: string): void {
    const source = networkMap.resources.find(service => service.id === sourceId)
    setSourceService(source)

    setRelation({ source: sourceId, target: destinationId, properties: { type: 'TCP', port: '' } })
  }

  return (
    <Layout.Horizontal width="100%" height="100%">
      <Container width="75%" height="100%">
        <>
          {nodes.length === 0 ? (
            <Layout.Vertical height="100%" flex={{ align: 'center-center' }} background={Color.WHITE} spacing="medium">
              <img src={noServiceIllustration} width={250} height={111} />
              <Text font={{ variation: FontVariation.H5 }}>{getString('discovery.noServiceSelected')}</Text>
              <Text width="40%">{getString('discovery.noServiceSelectedDescription')}</Text>
            </Layout.Vertical>
          ) : (
            <ReactFlowProvider>
              <NetworkGraph
                nodes={nodes}
                edges={edges}
                isNodeConnectable
                onNodeConnection={onNodeConnection}
                onNodeClick={node => setSourceService(node.data)}
              />
            </ReactFlowProvider>
          )}
          <Button
            className={css.backButton}
            variation={ButtonVariation.TERTIARY}
            text={getString('back')}
            onClick={() => handleTabChange(StudioTabs.SELECT_SERVICES)}
          />
        </>
      </Container>
      <Container width="35%" height="100%">
        {sourceService && (
          <NodeConnectionsSideBar
            sourceService={sourceService}
            networkMap={networkMap}
            updateNetworkMap={updateNetworkMap}
            newRelation={relation}
            closeSideBar={() => setSourceService(undefined)}
          />
        )}
        {!sourceService && <NoSelectionSideBar networkMap={networkMap} updateNetworkMap={updateNetworkMap} />}
      </Container>
    </Layout.Horizontal>
  )
}
