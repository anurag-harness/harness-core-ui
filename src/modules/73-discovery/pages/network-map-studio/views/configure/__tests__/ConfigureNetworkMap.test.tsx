/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { fireEvent, render } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import routes from '@common/RouteDefinitions'
import { accountPathProps, modulePathProps, networkMapPathProps, projectPathProps } from '@common/utils/routeUtils'
import { NetworkGraphProps } from '@modules/73-discovery/components/NetworkGraph/NetworkGraph'
import { mockEmptyNetworkMap, mockNetworkMap } from './mocks'
import ConfigureNetworkMap from '../ConfigureNetworkMap'

let callNodeFunctions = true

jest.mock('@discovery/components/NetworkGraph/NetworkGraph', () => ({
  ...jest.requireActual('@discovery/components/NetworkGraph/NetworkGraph'),
  __esModule: true,
  default: (props: NetworkGraphProps) => {
    if (callNodeFunctions) {
      callNodeFunctions = false
      props.onNodeConnection?.('64c229096fb4bc8feefc933b', '64c229096fb4bc8feefc933d')
      props.onNodeClick?.({
        width: 90,
        height: 97,
        position: {
          x: 12,
          y: 12
        },
        type: 'networkMapHexagon',
        expandParent: true,
        id: '64c229096fb4bc8feefc933b',
        data: {
          id: '64c229096fb4bc8feefc933b',
          kind: 'discoveredservice',
          name: 'cartservice',
          kubernetes: {
            namespace: 'boutique'
          }
        },
        parentNode: 'boutique',
        targetPosition: 'top' as any,
        sourcePosition: 'bottom' as any,
        style: {},
        selected: true,
        dragging: false,
        positionAbsolute: {
          x: 62,
          y: 62
        }
      })
    }
    return <div className={'networkGraph'}>Network Graph</div>
  }
}))

jest.mock('@discovery/pages//network-map-studio/views/configure/NodeConnectionsSideBar', () => ({
  ...jest.requireActual('@discovery/pages//network-map-studio/views/configure/NodeConnectionsSideBar'),
  __esModule: true,
  default: () => {
    return <div className="nodeConnectionsSideBar">NodeConnectionsSideBar</div>
  }
}))

const PATH = routes.toCreateNetworkMap({
  ...accountPathProps,
  ...projectPathProps,
  ...modulePathProps,
  ...networkMapPathProps
})
const PATH_PARAMS = {
  accountId: 'accountId',
  orgIdentifier: 'default',
  projectIdentifier: 'Discovery_Test',
  module: 'chaos',
  dAgentId: 'dAgent-1',
  networkMapId: 'testnetworkmap'
}

const handleTabChange = jest.fn()
const setNetworkMap = jest.fn().mockImplementation(() => mockNetworkMap)

describe('<ConfigureNetworkMap /> tests with data', () => {
  beforeEach(() => jest.clearAllMocks())

  test('should render component', async () => {
    const { container } = render(
      <TestWrapper path={PATH} pathParams={PATH_PARAMS}>
        <ConfigureNetworkMap
          networkMap={mockNetworkMap}
          updateNetworkMap={setNetworkMap}
          handleTabChange={handleTabChange}
        />
      </TestWrapper>
    )

    expect(container).toMatchSnapshot()
  })

  test('should press back button', async () => {
    const { container, getByText } = render(
      <TestWrapper path={PATH} pathParams={PATH_PARAMS}>
        <ConfigureNetworkMap
          networkMap={mockEmptyNetworkMap}
          updateNetworkMap={setNetworkMap}
          handleTabChange={handleTabChange}
        />
      </TestWrapper>
    )

    fireEvent.click(getByText('back'))

    expect(container).toMatchSnapshot()
  })
})
