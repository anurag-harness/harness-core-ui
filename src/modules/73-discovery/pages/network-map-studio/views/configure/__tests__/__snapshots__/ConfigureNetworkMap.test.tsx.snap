// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConfigureNetworkMap /> tests with data should press back button 1`] = `
<div>
  <div
    class="Layout--horizontal StyledProps--main"
    style="width: 100%; height: 100%;"
  >
    <div
      class="StyledProps--font StyledProps--main"
      style="width: 75%; height: 100%;"
    >
      <div
        class="Layout--vertical Layout--layout-spacing-medium StyledProps--main StyledProps--flex StyledProps--flex-align-center-center StyledProps--background-white"
        style="height: 100%;"
      >
        <img
          height="111"
          src="test-file-stub"
          width="250"
        />
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-variation-h5"
        >
          discovery.noServiceSelected
        </p>
        <p
          class="StyledProps--font StyledProps--main"
          style="width: 40%;"
        >
          discovery.noServiceSelectedDescription
        </p>
      </div>
      <button
        aria-label="back"
        class="bp3-button bp3-disabled bp3-loading Button--button StyledProps--font StyledProps--main backButton Button--with-current-color Button--variation Button--variation-tertiary"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <div
          class="bp3-spinner bp3-button-spinner"
        >
          <div
            class="bp3-spinner-animation"
          >
            <svg
              height="20"
              stroke-width="16.00"
              viewBox="-3.00 -3.00 106.00 106.00"
              width="20"
            >
              <path
                class="bp3-spinner-track"
                d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
              />
              <path
                class="bp3-spinner-head"
                d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
                pathLength="280"
                stroke-dasharray="280 280"
                stroke-dashoffset="210"
              />
            </svg>
          </div>
        </div>
        <span
          class="bp3-button-text"
        >
          back
        </span>
      </button>
    </div>
    <div
      class="StyledProps--font StyledProps--main"
      style="width: 35%; height: 100%;"
    >
      <div
        class="Layout--vertical Layout--layout-spacing-small StyledProps--main sidebar StyledProps--padding-large"
      >
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-variation-card-title"
        >
          discovery.addCustomizedConfigurations
        </p>
        <p
          class="StyledProps--font StyledProps--main"
        >
          discovery.addCustomizedConfigurationsDesc
        </p>
        <div
          class="Layout--vertical Layout--layout-spacing-medium StyledProps--main StyledProps--flex StyledProps--flex-alignItems-flex-start StyledProps--padding-top-medium"
        >
          <span
            class="bp3-popover-wrapper"
          >
            <span
              class="bp3-popover-target"
            >
              <button
                aria-label="discovery.newRelation"
                class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--withLeftIcon Button--variation Button--variation-link"
                type="button"
              >
                <span
                  class="bp3-icon bp3-icon-plus StyledProps--main StyledProps--padding-right-xsmall"
                  icon="plus"
                >
                  <svg
                    data-icon="plus"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <desc>
                      plus
                    </desc>
                    <path
                      d="M13 7H9V3c0-.55-.45-1-1-1s-1 .45-1 1v4H3c-.55 0-1 .45-1 1s.45 1 1 1h4v4c0 .55.45 1 1 1s1-.45 1-1V9h4c.55 0 1-.45 1-1s-.45-1-1-1z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
                <span
                  class="bp3-button-text"
                >
                  discovery.newRelation
                </span>
              </button>
            </span>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<ConfigureNetworkMap /> tests with data should render component 1`] = `
<div>
  <div
    class="Layout--horizontal StyledProps--main"
    style="width: 100%; height: 100%;"
  >
    <div
      class="StyledProps--font StyledProps--main"
      style="width: 75%; height: 100%;"
    >
      <div
        class="networkGraph"
      >
        Network Graph
      </div>
      <button
        aria-label="back"
        class="bp3-button Button--button StyledProps--font StyledProps--main backButton Button--with-current-color Button--variation Button--variation-tertiary"
        type="button"
      >
        <span
          class="bp3-button-text"
        >
          back
        </span>
      </button>
    </div>
    <div
      class="StyledProps--font StyledProps--main"
      style="width: 35%; height: 100%;"
    >
      <div
        class="nodeConnectionsSideBar"
      >
        NodeConnectionsSideBar
      </div>
    </div>
  </div>
</div>
`;
